### PR TITLE
Task runner

### PR DESCRIPTION
**Pull request by lonnieezell**
- Implemented `shouldRun()` on Tasks. 
- Implemented the basic logic of TaskRunner::run().

NOTE: Since I've noticed other frameworks having notes regarding how using Timezones for scheduled tasks was problematic because runs could be missed or executed twice thanks to Daylight Savings time, I removed the ability to specify a timezone.

### CODE REVIEW

**Review by MGatner — COMMENTED** [review]
I will have to look at this on desktop bc there is too much happening for mobile.  It for the tests you need to set your GitHub secret for COMPOSER_AUTH

File: docs/index.md [hunk 1]
```diff
@@ -114,7 +114,6 @@ There are a number of ways available to specify how often the task is called.
 | ->yearly('12:34am')           | Runs the first day of the year.                                       |
 | ->weekdays('1:23pm')          | Runs M-F at 12:00 am unless time passed in.                           |
 | ->weekends('2:34am')          | Runs Saturday and Sunday at 12:00 am unless time passed in.           |
-| ->timezone('America/Chicago') | Specifies a timezone to use when determine if it runs.                |
```
Comment (MGatner) [inline docs/index.md:117]: Somehow this messed up the table formatting.
Screen Shot 2020-12-09 at 10 07 38 AM

File: src/Task.php [hunk 2]
```diff
@@ -109,27 +115,34 @@ public function run()
...
+			$cron->testTime($testTime);
+		}
 
-		return $this;
+		// Make sure the timezone is updated
+		// if this task is specifying it.
+		if (! empty($this->timezone))
+		{
+			$cron->setTimezone($this->timezone);
+		}
```
Comment (MGatner) [inline src/Task.php]: So Tasks can still specify timezones even if the Scheduler cannot check them?

File: src/TaskRunner.php [hunk 3]
```diff
@@ -1,19 +1,51 @@
...
+				if ($task->shouldRun())
+				{
+					$task->run();
+				}
+			}
+		}
+		catch (\Throwable $e)
+		{
+			log_message('error', $e->getMessage(), $e->getTrace());
+		}
```
Comment (MGatner) [inline src/TaskRunner.php]: I think the `try...catch` block should be just around `$task->run()` so we don't accidentally hide errors in `shouldRun()`.
```suggestion
		foreach ($tasks as $task)
		{
			if ($task->shouldRun())
			{
				try
				{
					$task->run();
				}
				catch (\Throwable $e)
				{
					log_message('error', $e->getMessage(), $e->getTrace());
				}
			}
		}
```

**Review by MGatner — CHANGES_REQUESTED** [review]
Looks pretty good, a few comments.

**Comment by MGatner** [comment]
Too many changes for me to review on mobile - I'll check it out on desktop when I get a chance. Mention me when you're ready if I haven't gotten to it yet.

**Comment by lonnieezell** [comment]
@MGatner changes pushed, please review again. 

There are a couple of PHPStan's not passing - and I don't believe they're accurate. Within TaskRunner, for example, unless `Time::now()` throws an exception, which I don't believe it can do, the variables should always be present within the final block.

File: src/TaskRunner.php (hunk 3 above)
Comment (MGatner) [reply src/TaskRunner.php]: @lonnieezell  I would still like to see this change.

**Comment by MGatner** [comment]
I can pull this locally and take a look at the static analysis tomorrow. 

> Within TaskRunner, for example, unless Time::now() throws an exception, which I don't believe it can do, the variables should always be present within the final block.

`Time::now()` should be moved out of the `try` block then. `$output` will be conditional on the exception so I do think PHPStan is correct to flag that.

> spark

You have a whole copy of the **spark** file in this PR - accidental I assume??

**Comment by lonnieezell** [comment]
@MGatner updated once again.

**Comment by MGatner** [comment]
Looks good!

**Review by MGatner — APPROVED** [review]

**Pull request merged**